File: javascript/349.intersection-of-two-arrays.js
/*
 * @lc app=leetcode id=349 lang=javascript
 *
 * [349] Intersection of Two Arrays
 */

// @lc code=start
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[]}
 */
var intersection = function (nums1, nums2) {
  const result = [];
  const nums_set = new Set(nums1);
  for (const n of nums2) {
    if (nums_set.has(n)) {
      result.push(n);
      nums_set.delete(n);
    }
  }
  return result;
};
// @lc code=end

File: rust/349.intersection-of-two-arrays.rs
/*
 * @lc app=leetcode id=349 lang=rust
 *
 * [349] Intersection of Two Arrays
 */

// @lc code=start
impl Solution {
    pub fn intersection(nums1: Vec<i32>, nums2: Vec<i32>) -> Vec<i32> {
        use std::collections::HashSet;
        use std::iter::FromIterator;
        let mut nums_set: HashSet<i32> = HashSet::from_iter(nums1.into_iter());
        let mut result = Vec::new();
        for n in nums2 {
            if nums_set.contains(&n) {
                result.push(n);
                nums_set.remove(&n);
            }
        }
        return result;
    }
}
// @lc code=end
